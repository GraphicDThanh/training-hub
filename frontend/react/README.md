# React Redux Training
[Plan doc](https://docs.google.com/document/d/1lYywN0RIhrwsyFZ8uPZuaN5QM78mYuXFD608vS8YEbM/edit?pli=1#)
## ES6
## React Basic
### Practice 1 - examples React
## React Advanced
### Practice 2
## Redux Basic
## Redux Advanced
## Redux Recipes
