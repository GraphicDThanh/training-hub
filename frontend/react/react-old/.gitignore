node_modules
out
dist
.DS_Store
.git
.docpad.db
vendor/
.vagrant
bower-components/
bower_components/
npm-debug.log
.gitignore~
vagrant/Vagrantfile~
build/
