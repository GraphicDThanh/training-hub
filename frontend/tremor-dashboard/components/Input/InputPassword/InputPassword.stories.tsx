import type { Meta, StoryObj } from "@storybook/react";

// Components
import InputPassword from "./InputPassword";

const meta = {
  title: "Components/Common/InputPassword",
  component: InputPassword,
  tags: ["autodocs"],
  argTypes: {
    label: { description: "Label of input field" },
    additionalClass: {
      description: "Add more style for component",
    },
    placeholder: {
      description: "Show text placeholder",
    },
    errorMessage: {
      description: "Show error message when got error",
    },
  },
} as Meta<typeof InputPassword>;

export default meta;

type Story = StoryObj<typeof meta>;

export const InputPasswordDefault: Story = {
  args: {
    label: "Password",
  },
};

export const InputPasswordWithError: Story = {
  args: {
    label: "Password",
    value: "abcABC@131313",
    errorMessage: "Password is incorrect",
  },
};

export const InputPasswordDisabled: Story = {
  args: {
    label: "Password",
    value: "abcABC@131313",
    disabled: true,
  },
};
