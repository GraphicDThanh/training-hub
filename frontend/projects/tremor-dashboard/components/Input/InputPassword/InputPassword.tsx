"use client";

import { ChangeEvent, ReactNode, forwardRef, useState } from "react";

// Helpers
import { cn } from "@/helpers";

// Components
import { Text, Flex } from "@tremor/react";

// Icons
import { RiEyeFill, RiEyeOffFill } from "react-icons/ri";

interface InputPasswordProps
  extends React.InputHTMLAttributes<HTMLInputElement> {
  id: string;
  label?: string;
  errorMessage?: string;
  additionalClass?: string;
}

export type InputProps = Omit<InputPasswordProps, "stepper" | "onChange"> & {
  defaultValue?: string;
  icon?: ReactNode;
  errorMessage?: string;
  onChange: (value: string) => void;
};

const InputPassword = forwardRef<HTMLInputElement, InputProps>(
  (
    {
      id,
      label,
      placeholder,
      additionalClass = "",
      errorMessage = "",
      onChange,
      disabled = false,
      ...rest
    },
    ref,
  ) => {
    const [showPassword, setShowPassword] = useState(false);

    const toggleShowPassword = () => {
      setShowPassword(!showPassword);
    };

    const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
      const value: string =
        event.target.value.trimStart().replace(/\s+/g, " ") ?? "";

      onChange?.(value);
    };

    return (
      <div className="w-full">
        <label
          htmlFor={id}
          className="text-grey text-sm dark:text-lighter after:content-['*'] after:text-red-500 after:text-md">
          {label}
        </label>
        <Flex className="relative">
          <div className="w-full">
            <div className="relative">
              <input
                ref={ref}
                id={id}
                type={showPassword ? "text" : "password"}
                placeholder={placeholder}
                onChange={handleChange}
                disabled={disabled}
                className={cn(
                  "block py-1.5 px-0 w-full text-sm text-primary bg-transparent border-0 border-b-2 border-gray-300 appearance-none dark:text-white dark:border-gray-600 dark:focus:border-gray-500 focus:outline-none focus:ring-0 focus:border-gray-600 peer disabled:cursor-not-allowed disabled:opacity-50",
                  additionalClass,
                )}
                {...rest}
              />
              {!disabled && (
                <div
                  className="absolute right-0 top-0 mt-2 mr-2 hover:cursor-pointer hover:opacity-75"
                  onClick={toggleShowPassword}>
                  {showPassword ? (
                    <RiEyeOffFill className="text-primary text-xl dark:hover:text-white" />
                  ) : (
                    <RiEyeFill className="text-primary text-xl dark:hover:text-white" />
                  )}
                </div>
              )}
            </div>
            {errorMessage && (
              <Text className="pt-1 text-xs text-red-500 dark:text-red-500">
                {errorMessage}
              </Text>
            )}
          </div>
        </Flex>
      </div>
    );
  },
);

InputPassword.displayName = "InputPassword";

export default InputPassword;
