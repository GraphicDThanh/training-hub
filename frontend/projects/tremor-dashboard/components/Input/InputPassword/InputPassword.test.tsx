import { fireEvent, render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";

import InputPassword from "./InputPassword";

const mockProps = {
  id: "test-input-password",
  label: "Input Password",
  onChange: jest.fn(),
};

describe("Test InputPassword component", () => {
  it("Should match snapshot", () => {
    const { container } = render(<InputPassword {...mockProps} />);
    expect(container).toMatchSnapshot();
  });

  it("Should render InputPassword with error message", () => {
    render(
      <InputPassword errorMessage="Password is incorrect" {...mockProps} />,
    );
    const error = screen.queryByText(/Password is incorrect/i);
    expect(error).toBeInTheDocument();
  });

  it("Should call onChange handler with correct parameters", () => {
    const { getByLabelText } = render(<InputPassword {...mockProps} />);

    const inputElement = getByLabelText("Input Password");

    fireEvent.change(inputElement, { target: { value: "test value" } });

    expect(mockProps.onChange).toHaveBeenCalledTimes(1);
    expect(inputElement).toHaveValue("test value");
  });
});
