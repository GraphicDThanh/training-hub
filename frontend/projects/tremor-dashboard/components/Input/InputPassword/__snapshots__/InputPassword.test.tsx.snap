// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test InputPassword component Should match snapshot 1`] = `
<div>
  <div
    class="w-full"
  >
    <label
      class="text-grey text-sm dark:text-lighter after:content-['*'] after:text-red-500 after:text-md"
      for="test-input-password"
    >
      Input Password
    </label>
    <div
      class="tremor-Flex-root flex w-full flex-row justify-between items-center relative"
    >
      <div
        class="w-full"
      >
        <div
          class="relative"
        >
          <input
            class="block py-1.5 px-0 w-full text-sm text-primary bg-transparent border-0 border-b-2 border-gray-300 appearance-none dark:text-white dark:border-gray-600 dark:focus:border-gray-500 focus:outline-none focus:ring-0 focus:border-gray-600 peer disabled:cursor-not-allowed disabled:opacity-50"
            id="test-input-password"
            type="password"
          />
          <div
            class="absolute right-0 top-0 mt-2 mr-2 hover:cursor-pointer hover:opacity-75"
          >
            <svg
              class="text-primary text-xl dark:hover:text-white"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M1.18164 12C2.12215 6.87976 6.60812 3 12.0003 3C17.3924 3 21.8784 6.87976 22.8189 12C21.8784 17.1202 17.3924 21 12.0003 21C6.60812 21 2.12215 17.1202 1.18164 12ZM12.0003 17C14.7617 17 17.0003 14.7614 17.0003 12C17.0003 9.23858 14.7617 7 12.0003 7C9.23884 7 7.00026 9.23858 7.00026 12C7.00026 14.7614 9.23884 17 12.0003 17ZM12.0003 15C10.3434 15 9.00026 13.6569 9.00026 12C9.00026 10.3431 10.3434 9 12.0003 9C13.6571 9 15.0003 10.3431 15.0003 12C15.0003 13.6569 13.6571 15 12.0003 15Z"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
